[IMP] edi: Detection of Pending Manual EDIs and adding it to form view

etobella opened the pull request:
This PR adds a hook on account.move for manual generation of EDIs. Another option is to add directly on EDI Mixin :smile: 

@simahawk

simahawk commented on edi_account/models/account_move.py:
```diff
@@ -29,3 +46,9 @@ def action_invoice_paid(self):
+
+    def action_generate_edi(self):
+        """This launches a notification for generating missing manual EDIs"""
+        self._event("on_generate_account_edi").notify(
```
This is going to trigger the event even if you have not records.
Also what would be the behavior?
Maybe we can discuss how to do it generally...

simahawk replied:
As discussed in call, we could have:
1. model_ids on exc type
2. enable_domain to filter records matching the type
3. enable_snippet here you could do what you want and return true/false when it applies or not to the current record

The next step would be: automatically show buttons depending on this based on a serialized computed field that tells you per exchange type which btn show.

etobella commented:
@simahawk First approach of what we talked about. Stil pending some tests and some extra documentation :smile:

etobella commented:
Tests have been improved, the error on runbot is solved on #268

etobella commented:
![image](https://user-images.githubusercontent.com/28590170/101163240-aa1a1480-3633-11eb-9f9b-5c405d520508.png)

This is how it will be shown :smile:

jarroyomorales reviewed (approved):
Tested and works as expected! :+1:

simahawk commented:
@etobella seems amazing! Thanks! I'll review later or tomorrow.

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,104 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+        exchange_types = (
+            self.env["edi.exchange.type"]
+            .sudo()
+            .search([("model_ids.model", "=", self._name), ("backend_id", "!=", False)])
```
why backend_id False? This excludes all types that are not assigned to a specific backend but only to a backend type.

etobella replied:
Because I need a backend for generating the exchange record and I only have the record and the exchange type :thinking:

simahawk replied:
that's why we should have a wizard IMO to select a backend or auto assign it IF there's only one backend for that specific type.backend_type_id.

etobella replied:
True, I will work on it :wink:

etobella replied:
Done

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,104 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+            .search([("model_ids.model", "=", self._name), ("backend_id", "!=", False)])
+        )
+        result = {}
+        for exchange_type in exchange_types:
+            if exchange_type.enable_domain and not self.search(
+                [("id", "=", self.id)]
+                + safe_eval(exchange_type.enable_domain, self._get_eval_context())
+            ):
+                continue
+            if exchange_type.enable_snippet:
+                eval_ctx = {
+                    "record": self,
+                    "exchange_type": exchange_type,
+                }
+                eval_ctx.update(self._get_eval_context())
+                safe_eval(
+                    exchange_type.enable_snippet, eval_ctx, mode="exec", nocopy=True
+                )
+                if not eval_ctx.get("result", False):
+                    continue
+            result[exchange_type.id] = exchange_type.display_name
```
do not search for the record, you have it.  Use `filtered_domain` to apply the domain on it:
```suggestion
        for exchange_type in exchange_types:
            eval_ctx = dict(
                self._get_eval_context(),
                record=self,
                exchange_type=exchange_type
            )
            domain = safe_eval(
                exchange_type.enable_domain or "[]", eval_ctx
            )
            if not self.filtered_domain(domain):
                continue
            if exchange_type.enable_snippet:
                safe_eval(
                    exchange_type.enable_snippet, eval_ctx, mode="exec", nocopy=True
                )
                if not eval_ctx.get("result", False):
                    continue
            result[exchange_type.id] = exchange_type.display_name
```

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,104 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
     )
     exchange_record_count = fields.Integer(compute="_compute_exchange_record_count")
+    expected_edi_configuration = fields.Serialized(
+        compute="_compute_expected_edi_configuration"
```
```suggestion
        compute="_compute_expected_edi_configuration"
        default={}
```

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,104 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+        # TODO: Define which key the component should have
+        if component:
+            # TODO: Define the function name properly
+            component.manual_generation()
```
why another component? What's manual? I think the action should depend on what is missing:
1. missing a record? create it
2. missing data? Receive it
3. missing send? send it
etc...

simahawk replied:
All in all it's not clear what's the purpose of this method :thinking: :sweat_smile:

etobella replied:
Yes, you are right, I will add an event just in case and keeping it the simplest way :wink:

simahawk commented on edi/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,26 @@
+This module is intended to be inherited by other modules in order to define the proper
```
yes and no... the aim is also to allow to not have to code anything and just work w/ configuration

simahawk commented on edi/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,26 @@
+* Backend
+* Connectors
+
+Connector definition
```
Connector != component :wink:

simahawk commented on edi/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,26 @@
+* kind can be: `generate`, `send`, `check`, `process`, `receive`
+* code is the `{backend type code}` or `{backend type code}.{exchange type code}`
+
+Manual EDI generation
```
To be clarified

simahawk commented on edi/readme/USAGE.rst:
```diff
@@ -0,0 +1,31 @@
+* Generation of data
+* Checking of data
+* Sending data
+* Reviewing if the data was processed properly by the other party
```
Validation

simahawk commented on edi/readme/USAGE.rst:
```diff
@@ -0,0 +1,31 @@
+
+* Creation
+* Reception of data
+* Checking data
```
Validation

simahawk commented on edi/tests/test_consumer_mixin.py:
```diff
@@ -7,13 +7,15 @@
 # If you still want to run `edi` tests w/ pytest when this happens, set this env var.
 @unittest.skipIf(os.getenv("SKIP_EDI_CONSUMER_CASE"), "Consumer test case disabled.")
-class TestConsumerMixinCase(EDIBackendCommonTestCase):
+class TestConsumerMixinCase(EDIBackendCommonComponentRegistryTestCase):
```
the component savepoint case is enough as you don't register any new component on the fly

simahawk commented on edi/tests/test_consumer_mixin.py:
```diff
@@ -49,3 +51,50 @@ def test_mixin(self):
+                "model_ids": [
+                    (
+                        4,
+                        self.env["ir.model"]
+                        .search([("model", "=", self.consumer_record._name)], limit=1)
+                        .id,
```
```suggestion
                        self.env['ir.model']._get_id(self.consumer_record._name),
```

simahawk commented on edi/tests/test_consumer_mixin.py:
```diff
@@ -49,3 +51,50 @@ def test_mixin(self):
+        )
+
+    def test_form(self):
+        """Testing that the form has inherited the fields and inserted them
```
I would test the the form xml instead :wink:

etobella commented:
@simahawk comments attended

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,96 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+            "res_id": self.id,
+        }
+
+    def generate_edi_configuration(self, exchange_type_id):
```
this should be named `edi_create_exchange_record` and their related methods renamed accordingly.

etobella replied:
Done

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,6 +25,96 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+        exchange_record = backend.create_record(
+            exchange_type.code, self._generate_edi_configuration_vals(exchange_type)
+        )
+        self._event("on_generate_manual_edi").notify(self, exchange_record)
```
I don't think you need this event: you can already hook to `on_create` no?
What's the goal you had in mind?

etobella replied:
The goal of this listener is to allow configuration to process the record if needed (it depends on the listener). Why not using the on_create, because the on_create does not know why the record was generated, so it is always better to reuse it (different behavior on standard creation and manual creation?)

simahawk replied:
ok, call it `on_edi_generate_manual`

etobella commented:
@simahawk Comments attended or answered :wink:

simahawk commented on edi/models/edi_exchange_consumer_mixin.py:
```diff
@@ -19,13 +25,114 @@ class EDIExchangeConsumerMixin(models.AbstractModel):
+    def edi_create_exchange_record(self, exchange_type_id):
+        self.ensure_one()
+        exchange_type = self.env["edi.exchange.type"].browse(exchange_type_id)
+        if exchange_type.backend_id:
```
you can do the same if you find only one backend for exchange_type.backend_type_id :wink:

etobella replied:
OK

simahawk commented on edi/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1,28 @@
+Manual EDI generation
+~~~~~~~~~~~~~~~~~~~~~
+
+Also, the Exchange can be accessed on debug mode and edit `model_ids`, `enable_domain`
```
this should be rephrased...

simahawk commented on edi/views/edi_exchange_type_views.xml:
```diff
@@ -35,6 +35,15 @@
                             <field name="exchange_file_auto_generate" />
                             <field name="ack_type_id" />
                         </group>
+                        <group name="manual" group="base.group_no_one">
```
when #275  is merged we can use the specific group for advanced settings I added there :wink:

etobella replied:
OK

etobella commented:
Comments attended, still missing some tests for increasing coverage

etobella committed:
[IMP] edi: Improve readme

etobella committed:
[IMP] edi: Allow to manually generate EDI Exchange records

etobella commented:
@simahawk Comments attended and tests added

etobella commented:
Rebased and fixed conflicts, I think that the main issue here is the readme. If you want, I can remove it from here and create a PR just for this

@simahawk

simahawk commented:
> Rebased and fixed conflicts, I think that the main issue here is the readme. If you want, I can remove it from here and create a PR just for this
> 
> @simahawk

yes please :pray:

etobella commented:
Done

simahawk commented on edi/wizards/edi_exchange_record_create.py:
```diff
@@ -0,0 +1,24 @@
+    backend_type_id = fields.Many2one(
+        "edi.backend.type", related="exchange_type_id.backend_type_id"
+    )
+    backend_id = fields.Many2one("edi.backend", reaquired=True)
```
```suggestion
    backend_id = fields.Many2one("edi.backend", required=True)
```

etobella replied:
fixed

simahawk reviewed (approved):
small typo

simahawk commented:
/ocabot merge minor
